### ISSUE TITLE
Bug: MIDI on Daisy unstable

### ISSUE THREAD

**Issue by dromer** [Wasted-Audio/hvcc#175]
Several reports that MIDI crashes the E-S Daisy after a while.

No idea why yet.

**Comment by perkowitz** [comment]
For reference, some discussion here: https://forum.electro-smith.com/t/difficulties-using-pure-data-daisy-patch-midi/4995

And I'm happy to test possible fixes, if that would help.

**Comment by dromer** [comment]
What we need is someone that understands the problem and that can debug the issue :)

**Comment by dromer** [comment]
@perkowitz today I merged a "fix" that more properly sets up the messagePool in the constructor.

If you are able to run hvcc from the commandline then it's appreciated if you would give it a test :)

I ran a polyphonic synth patch, receiving polyphony midi notes, for something like 2.5 hrs without any crashes.
(although to be honest without this specific fix the patch also ran for like 1.5hrs ..)

**Comment by perkowitz** [comment]
I've mainly been using it from plugdata -- is there a way to build the toolchain from this repo and just slot it into my plugdata installation? 

I'm not having much luck so far with hvcc directly -- no matter how many times pip tells me it's installed, I can't find the binary anywhere.

**Comment by dromer** [comment]
> is there a way to build the toolchain from this repo and just slot it into my plugdata installation?

Yes, but that's quite advanced and I have no idea how to do this on non-Linux systems :shrug:

**Comment by perkowitz** [comment]
I haven't been able to test it yet, because I'm finding MIDI isn't working at all in my Pd patches on the daisy  ¯\_(ツ)_/¯

**Comment by perkowitz** [comment]
Okay, I have midi patches working again, but only with plugdata. I'm working on swapping in a new heavy toolchain so I can test it from plugdata

**Comment by dromer** [comment]
@perkowitz if you were using pd2dsy: this uses a very old version of hvcc and does not get any updates with new features. so it is expected that certain things do not work (or ever will).

**Comment by perkowitz** [comment]
I've been mainly using plugdata, but I thought it might be easier to update with pd2dsy, but clearly not. Unfortunately, I haven't yet figured out why plugdata fails when I swap the hvcc folder in plugdata's toolchain folder.

**Comment by perkowitz** [comment]
I have been unable to get plugdata to recognize an updated hvcc toolchain. I've also been unable to build hvcc from the repo and run it, on either mac or linux. I would like to test the fix but can't so far.

**Comment by andybeta** [comment]
I'm also having this problem with the Daisy Pod, even with even a very stripped down patch. My midi in is coming from an Akai MPK mini which afaik has no clock signal (although I might wrong.) Same behaviour from my ep-133 which defintely does have clock. 

CTLIN will accelerate the crash time, but it still happens with pure midi input.

Pollen(8) handles a ton more functionality than I can write and has never crashed on midi in for me, so I'm assuming the problem is not with the Daisy itself.

I need to look into rebuilding an updated toolchain, but windows is hampering me. I might setup a linux partition just to test this.

> Yes, but that's quite advanced and I have no idea how to do this on non-Linux systems 🤷

Do you have a link handy to do this on Linux please @dromer?

**Comment by dromer** [comment]
@andybeta you can find the toolchain here: https://github.com/plugdata-team/plugdata-heavy-toolchain/

**Comment by stephenhensley** [comment]
I've finally had a little bit of time to look into this, and might have found a (pleasantly simple) fix.

I was able to recreate the issue with a basic MIDI patch.
I attached a debugger to see what was triggering the hard fault (since thats what it _sounded_ like was happening). However, it looks like the `HeavyContext::sendMessageToReceiver` was getting stalled on: `HV_SPINLOCK_ACQUIRE(inQueueLock);`

So from that, and the call stack, it looks like what might have been happening was:

The main while() loop in the target application would call it's MidiHandler function. 
During this function, a MIDI message would end up calling the `sendMessageToReceiver`, acquing a lock on the atomic `inQueueLock`.
After the lock had been acquired, but before it was released, the audio interface IRQ would fire, causing execution to immediately jump to the audio callback. One of the first things done in the call back is call, `sendBangToReceiver` which requires acquiring the lock. Since the MIDI handling function will not resume until the audio callback returns, everything stalls indefinitely.

Fortunately, it looks like the fix for this is very simple. libDaisy has a `ScopedIrqBlocker` class, that will disable interrupts as part of its constructor, and then re-enable them when it is deconstructed.

I took a minimal program that received MIDI and would crash in a reasonably short amount of time, and then added the following modification to the exported source code:

```cpp
void HandleMidiMessage(MidiEvent m)
{
  ScopedIrqBlocker block; //< Disables interrupts while in scope 
  for (int i = 0; i <= 2; ++i)
  {
    hv.sendMessageToReceiverV(HV_HASH_MIDIIN, 0, "ff",
                              (float)m.data[i],
                              (float)m.channel);
  }
  . . .
}
```

With this patched into the file, I have not experienced further instabilities.
Hopefully, that will be all that's required to resolve the issue 😄

**Comment by dromer** [comment]
Yay! hopefully we can get @perkowitz and @andybeta to test this as well ;)

I've also added it to the other functions as you recommended on Discord.

**Comment by ruffleFPV** [comment]
@dromer I can confirm it fixes the issue.

**Comment by dromer** [comment]
@ruffleFPV yay, thank you for confirming :)
